staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_strategy.go

```diff
@@ -0,0 +1,217 @@
+			ctx:        ctx,
+			client:     client,
+			verb:       verb,
+			resource:   resource,
+			guards:     append(preconditionClient.Guards(), transactionGuards...),
+			operations: ops,
+		}
+		txnResponse, txnError = gTxn.Then(operation).Commit()
+		if txnError != nil {
+			// errors transacting or failed guards terminate this transaction
+			return nil, txnError
+		}
+		if txnResponse.Succeeded {
+			// successful transaction, we're done
+			break
+		}
+		// no errors during the transaction, but we failed some guards - retry, using fresh data from the transaction
+	}
+
+	return txnResponse, nil
+}
+
+// contextKey allows for unique context value keys, as this type is unexported from this package
+type contextKey string
+
+const (
+	// contextKeyCache is the key under which we store a potentially cached object and metadata
+	contextKeyCache = contextKey("cached")
+)
+
+// PotentiallyCachedResponse is a response from etcd which may or may not be stale
+type PotentiallyCachedResponse struct {
+	response *etcdclient.GetResponse
+	cached      bool
+	needsUpdate bool
+}
+
+// preconditionKVClient is a client that knows how to read values from etcd and ensure that they are up-to-date
+// when future transactions are done assuming those values.
+type preconditionKVClient struct {
+	delegate *etcdclient.Client
+
+	guards []TransactionGuard
+}
+
+// Get serves potentially cached data, ensuring that all kvs exposed to the precondition calling this are recorded
+// in modRevision guards so that we know the data used to evaluate the precondition does not change before the
+// transaction is committed.
+func (p *preconditionKVClient) Get(ctx context.Context, key string, opts ...etcdclient.OpOption) (*etcdclient.GetResponse, error) {
+	item := ctx.Value(contextKeyCache)
+	if item == nil {
+		return nil, storage.NewInternalErrorf("found no cached item under key %q in context", contextKeyCache)
+	}
+	cachedResponse, ok := item.(*PotentiallyCachedResponse)
+	if !ok {
+		return nil, storage.NewInternalErrorf("found invalid cached item %T under key %q in context", item, contextKeyCache)
+	}
+	// if we don't have a cached response to return, or we need to update our cached version, fetch from etcd
+	if cachedResponse.response == nil || (cachedResponse.cached && cachedResponse.needsUpdate) {
+		response, err := p.delegate.Get(ctx, key, opts...)
+		if err != nil {
+			return response, err
+		}
+		(*cachedResponse).response = response
+		(*cachedResponse).cached = false
+		(*cachedResponse).needsUpdate = false
+	}
+	for i, kv := range cachedResponse.response.Kvs {
+		p.guards = append(p.guards, &AtModRevision{
+			key:         string(kv.Key),
+			modRevision: kv.ModRevision,
+			consumeUpdatedData: func(response *etcdclient.GetResponse) error {
+				if len(response.Kvs) != 1 {
+					return storage.NewInternalErrorf("invalid gRPC response from etcd: expected 1 response, got %d", len(response.Kvs))
+				}
+				(*cachedResponse).response.Header = response.Header
```

Comment:
Why are the response kvs aggregated but the `response.Header` and `cached` fields set on every iteration?

Reply:
This is probably a poor factoring - in the case where one call to `Get()` returned multiple `kv`s during the precondition, we need to assert that every single one of them did not change, individually, in the transaction guards; but we could re-issue the same `Get()` in the transaction to refresh the full response, instead of doing a one-by-one thing like this. Would be a little more tricky to get right but possible.